docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+    compressionDisabled: Boolean?,
+    compressionThreshold: Int?,
+    private val requestedCompressionAlgorithms: List<String>,
+) : HttpInterceptor {
+
+    // Set defaults for settings if they are not set
+    private val compressionDisabled = compressionDisabled ?: false
+    private val compressionThreshold = compressionThreshold ?: 10240
+
+    // List of supported compression algorithms
+    private val supportedCompressionAlgorithms: List<CompressionAlgorithm> = listOf(
+        Gzip()
+    )
+
+    override suspend fun modifyBeforeRetryLoop(
+        context: ProtocolRequestInterceptorContext<Any, HttpRequest>
+    ): HttpRequest {
+
+        // Verify min compression size setting is in range
+        val compressionThresholdAcceptableRange = 0..10485760
+        if (!compressionThresholdAcceptableRange.contains(compressionThreshold))
+            throw Exception("SDK config property 'requestMinCompressionSizeBytes' must be a value between 0 and 10485760 (inclusive)")
+
+        // Determine if going forward with compression
+        val payloadSize = context.protocolRequest.body.contentLength
+        val streamingPayload = payloadSize == null
+        if (!compressionDisabled && (streamingPayload || payloadSize!! >= compressionThreshold)) {
+
+            // Check if requested algorithm(s) is supported
+            findAlgorithm()?.let { algorithm ->
+
+                // Attempt compression
+                val request = context.protocolRequest.toBuilder()
+                request.body = request.body
+                // TODO: Add compression for streams
+                addHeader(request, algorithm.algorithmId())
+                return request.build()
+            }
+        }
+        return context.protocolRequest
+    }
+}
+```
+
+Supporting function to help find if requested algorithm is supported:
+```kotlin
+/**
+ * Finds first match in supported and request
+ */
+private fun findAlgorithm() : CompressionAlgorithm? {
+    requestedCompressionAlgorithms.forEach { requestedAlgorithmAlgorithmId ->
+        supportedCompressionAlgorithms.forEach { supportedAlgorithm ->
+            if (supportedAlgorithm.algorithmId() == requestedAlgorithmAlgorithmId) return supportedAlgorithm
+        }
+    }
+    return null
+}
+
+// TODO: Return list instead of single algorithm
+
+```
+
+Supporting function to add header or modify header once compression is successful:
+```kotlin
+/**
+ * Appends the algorithm id to the content encoding header. Doesn't remove old content encodings if already present
+ * in header
+ */
+private fun addHeader(request: HttpRequestBuilder, algorithmId: String) {
+    val previousEncodings = request.headers["Content-Encoding"]
+    val contentEncodingHeaderPrefix = previousEncodings?.let { "$previousEncodings, " } ?: ""
+
+    request.headers.remove("Content-Encoding")
+    request.header(
+        "Content-Encoding",
+        "$contentEncodingHeaderPrefix${algorithmId}"
+    )
+}
```

Comment:
**Style**: Can be simplified:

```kotlin
private fun addHeader(request: HttpRequestBuilder, algorithmId: String) {
    request.headers.append("Content-Encoding", algorithmId)
}
```

You may not need a helper function for this.